tools/generate_schema_wrapper.py

```diff
@@ -702,82 +761,146 @@ def generate_vegalite_channel_wrappers(
 
-    return imports, "\n".join(code)
+    return MARK_MIXIN.format(methods="\n".join(code))
 
 
-def generate_vegalite_config_mixin(schemafile: Path) -> tuple[list[str], str]:
-    imports = [
-        "from . import core",
-        "from altair.utils import use_signature",
-    ]
+def _signature_args(
+    args: Iterable[str],
+    props: SchemaProperties,
+    *,
+    kind: Literal["method", "typed_dict"] = "method",
+) -> Iterator[str]:
+    """Lazily build a typed argument list."""
+    if kind == "method":
+        yield "self"
+        for p in args:
+            yield f"{p}: {props[p].to_type_repr(target='annotation', use_undefined=True)} = Undefined"
+        yield "**kwds"
+    elif kind == "typed_dict":
+        for p in args:
+            yield f"{p}: {props[p].to_type_repr(target='annotation', use_concrete=True)}"
+    else:
+        raise NotImplementedError
+
+
+def generate_mark_args(
+    info: SchemaInfo,
+) -> dict[Literal["method_args", "dict_args"], str]:
+    arg_info = codegen.get_args(info)
+    args = sorted((arg_info.required | arg_info.kwds) - {"type"})
+    dict_args = (f"{p}={p}" for p in args)
+    return {
+        "method_args": ", ".join(_signature_args(args, info.properties)),
+        "dict_args": ", ".join(chain(dict_args, ("**kwds",))),
+    }
+
+
+def generate_typed_dict_args(prop_info: SchemaInfo) -> str:
+    args = codegen.get_args(prop_info).required_kwds
+    it = _signature_args(args, prop_info.properties, kind="typed_dict")
+    return "\n    ".join(it)
+
+
+def generate_config_typed_dicts(fp: Path, /) -> Iterator[str]:
+    """TODO - Tidy up and use consistent naming."""
+    schema = load_schema(fp)
+    config = SchemaInfo({"$ref": "#/definitions/Config"}, rootschema=schema)
+    top_dict_annotations: list[str] = []
+    sub_dicts: dict[str, str] = {}
+    # FIXME: Replace with a recursive/graph approach
+    MANUAL_DEFS = (
+        "ScaleInvalidDataConfig",
+        "OverlayMarkDef",
+        "LinearGradient",
+        "RadialGradient",
+        "GradientStop",
+        "TooltipContent",
+        "DateTime",
+        "TimeIntervalStep",
+        "IntervalSelectionConfigWithoutType",
+        "PointSelectionConfigWithoutType",
+        "Feature<Geometry,GeoJsonProperties>",
+        "GeoJsonFeature",
+        "GeoJsonFeatureCollection",
+        "GeometryCollection",
+        "Point",
+        "Polygon",
+        "LineString",
+        "MultiPoint",
+        "MultiPolygon",
+        "MultiLineString",
+        "BrushConfig",
+        "MergedStream",
+        "DerivedStream",
+        "AutoSizeParams",
+        "Locale",
+        "VariableParameter",
+        "TopLevelSelectionParameter",
+        "PointSelectionConfig",
+        "IntervalSelectionConfig",
+        "BindInput",
+        "BindRange",
+        "BindDirect",
+        "BindCheckbox",
+        "BindRadioSelect",
+        "NumberLocale",
+        "TimeLocale",
+        "LegendStreamBinding",
+    )
```

Comment:
FIXME: Generate this selection, instead of manually defining

Reply:
This is how far I got, maybe it helps.

```python
import itertools as it
import json

from pathlib import Path
from typing import Any

from tools.schemapi.utils import SchemaInfo, get_valid_identifier

def load_schema(fp: Path, /) -> dict[str, Any]:
    """Reads and returns the root schema from ``fp``."""
    with fp.open(encoding="utf8") as f:
        root_schema = json.load(f)
    return root_schema


schema = load_schema(Path("altair/vegalite/v5/schema/vega-lite-schema.json"))
config = SchemaInfo({"$ref": "#/definitions/Config"}, rootschema=schema)


def is_relevant_object(schema: SchemaInfo) -> bool:
    return (
        schema.is_object()
        and not schema.is_value()
        and schema.title.strip()
        and schema.properties
    )


def get_all_objects(
    schema: SchemaInfo, recursion_depth: int = 0
) -> dict[str, SchemaInfo]:
    objects: dict[str, SchemaInfo] = {}
    if is_relevant_object(schema):
        objects[schema.title] = schema

    if recursion_depth < 10:
        for prop_info in it.chain(
            schema.properties.values(),
            schema.anyOf,
            schema.allOf,
            schema.oneOf,
            [schema.child(schema.items)] if schema.items else [],
        ):
            objects.update(
                get_all_objects(prop_info, recursion_depth=recursion_depth + 1)
            )
    return objects
    
    
all_objects = get_all_objects(config)
def_names = list(all_objects.keys())
MANUAL_DEFS = (
        "ScaleInvalidDataConfig",
        "OverlayMarkDef",
        "LinearGradient",
        "RadialGradient",
        "GradientStop",
        "TooltipContent",
        "DateTime",
        "TimeIntervalStep",
        "IntervalSelectionConfigWithoutType",
        "PointSelectionConfigWithoutType",
        "Feature<Geometry,GeoJsonProperties>",
        "GeoJsonFeature",
        "GeoJsonFeatureCollection",
        "GeometryCollection",
        "Point",
        "Polygon",
        "LineString",
        "MultiPoint",
        "MultiPolygon",
        "MultiLineString",
        "BrushConfig",
        "MergedStream",
        "DerivedStream",
        "AutoSizeParams",
        "Locale",
        "VariableParameter",
        "TopLevelSelectionParameter",
        "PointSelectionConfig",
        "IntervalSelectionConfig",
        "BindInput",
        "BindRange",
        "BindDirect",
        "BindCheckbox",
        "BindRadioSelect",
        "NumberLocale",
        "TimeLocale",
        "LegendStreamBinding",
    )
MANUAL_DEFS_VALID = [get_valid_identifier(k) for k in MANUAL_DEFS]
assert all(m in def_names for m in MANUAL_DEFS_VALID)  # Works
set(MANUAL_DEFS_VALID).symmetric_difference(def_names)  # def_names also includes the `Config` entries such as `AxisConfig` and some additional ones.
```

Some observations:
* The rules for reading out the relevant schema definitions is a bit wonky. I just tweaked it until I had what I wanted.
  * The recursion depth limit is pretty random but I had to set something. Unclear if there is a better way to stop earlier?
* It also includes predicate classes (`FieldRangePredicate`, ...) which are not present in your current `_config.py`. I'm not sure if they should be excluded or if they are relevant.
* The typed dicts should probably also accept strings (or dicts?) wherever a `ExprRef` is accepted. For example `AxisConfig.aria`.

Reply:
> This is how far I got, maybe it helps.
> ...

Will need to look into [this](https://github.com/vega/altair/pull/3536#discussion_r1743942369) in a bit more depth soon, but at first glance this is great thank you for the help @binste!

> It also includes predicate classes (`FieldRangePredicate`, ...) which are not present in your current `_config.py`. I'm not sure if they should be excluded or if they are relevant.

IIRC, I think I wouldn't have wanted that in there - but looking at `MANUAL_DEFs` again now maybe there are others that make less sense to apply in a theme?

Essentially I was trying to create a `TypedDict` if:
- After resolving references starting at `Config`, it would appear in some definition
- **And** it would be resonable to set a default for a property

All of the `...Config` classes are obvious choices, but some others I'm less sure on.

---

Regardless - really appreciate your time on this!

Reply:
# Update
Just tested this out and I can work with this @binste - thank you!

Hopefully will make some more progress tomorrow

Reply:
Awesome! Looking forward to it :) I'll definitely rewrite my own theme code with `ThemeConfig` once we release it.

Reply:
I haven't fully tested this but it might be easier to just go through all definitions in the top-level to get a mapping of types to the ones which it uses directly, `graph` in the example below. Next step would be to start with Config and look up children and always draw the new descendants from `graph`. Maybe it's the same, thought I just share it in case it sparks any ideas for you.

```python
from pathlib import Path
from typing import Any
import json
from tools.schemapi.utils import SchemaInfo

def load_schema(fp: Path, /) -> dict[str, Any]:
    """Reads and returns the root schema from ``fp``."""
    with fp.open(encoding="utf8") as f:
        root_schema = json.load(f)
    return root_schema


rootschema = load_schema(Path("altair/vegalite/v5/schema/vega-lite-schema.json"))

definitions: dict[str, SchemaInfo] = {}

for name in rootschema["definitions"]:
    schema = SchemaInfo({"$ref": "#/definitions/" + name}, rootschema=rootschema)
    definitions[name] = schema

graph: dict[str, list[str]] = {}

import itertools as it
for name, schema in definitions.items():
    graph[name] = []
    
    for prop_info in it.chain(
            schema.properties.values(),
            schema.anyOf,
            schema.allOf,
            schema.oneOf,
            [schema.child(schema.items)] if schema.items else [],
        ):
        if prop_info.refname:
            graph[name].append(prop_info.refname)
```

Reply:
Maybe this is easier, it doesn't run into any recursion issue and gives you a dictionary `graph` which you can walk. You could start from `Config` and then go down from there:

```python
from pathlib import Path
from typing import Any
import json
from tools.schemapi.utils import SchemaInfo

def load_schema(fp: Path, /) -> dict[str, Any]:
    """Reads and returns the root schema from ``fp``."""
    with fp.open(encoding="utf8") as f:
        root_schema = json.load(f)
    return root_schema


rootschema = load_schema(Path("altair/vegalite/v5/schema/vega-lite-schema.json"))

definitions: dict[str, SchemaInfo] = {}

for name in rootschema["definitions"]:
    schema = SchemaInfo({"$ref": "#/definitions/" + name}, rootschema=rootschema)
    definitions[name] = schema

# Key is the name of a definition and the value is a set of names of definitions
# that are children of the key definition.
graph: dict[str, set[str]] = {}

import itertools as it

def get_children(schema_info: SchemaInfo) -> set[str]:
    children: set[str] = set()
    
    for prop_info in it.chain(
            schema_info.properties.values(),
            schema_info.anyOf,
            [schema_info.child(schema_info.items)] if schema_info.items else [],
        ):
        if prop_info.refname and prop_info.is_reference():
            children.add(prop_info.refname)
        else:
            children.update(get_children(prop_info))
            
    return children
    

for name, schema in definitions.items():
    graph[name] = get_children(schema)       
```

I'm still double-checking if this captures everything but so far it looks promising :) Unrelated to this PR, I'm trying to analyze how types are related in general and how we could simplify this in Vega-Lite.

Reply:
@binste 

Adapting your original code, I've managed to select the schemas I wanted with:

```py
import keyword
def is_relevant_object(schema: SchemaInfo) -> bool:
    EXCLUDE = {"ExprRef", "RelativeBandSize", "ParameterPredicate"}
    return (
        schema.ref
        and schema.refname not in EXCLUDE
        and schema.properties
        and schema.type == "object"
        and not schema.is_value()
        and "field" not in schema.required
        and not (keyword.iskeyword(next(iter(schema.required), "")))
    )
```

Can you try plugging that in as a filter somewhere in your non-recursive version? https://github.com/vega/altair/pull/3536#discussion_r1748059188

Reply:
If you want, you can paste your code here and I'll add it directly in there as I'm not sure how it's most useful to you/what the output is you need. Else, you can try to replace `if prop_info.refname and prop_info.is_reference()` in my example code with the latest `is_relevant_object`

Reply:
> If you want, you can paste your code here and I'll add it directly in there as I'm not sure how it's most useful to you/what the output is you need.

<details>
<summary>Full (old) code</summary>

```py
import keyword
from itertools import chain

from tools import generate_schema_wrapper as gsw
from tools.schemapi.utils import SchemaInfo

schema = gsw.load_schema(Path("altair/vegalite/v5/schema/vega-lite-schema.json"))
config = SchemaInfo({"$ref": "#/definitions/Config"}, rootschema=schema)


def is_relevant_object(schema: SchemaInfo):
    EXCLUDE = {"ExprRef", "RelativeBandSize", "ParameterPredicate"}
    return (
        schema.ref
        and schema.refname not in EXCLUDE
        and schema.properties
        and schema.type == "object"
        and not schema.is_value()
        and "field" not in schema.required
        and not (keyword.iskeyword(next(iter(schema.required), "")))
    )


def get_all_objects(
    schema: SchemaInfo, recursion_depth: int = 0
) -> dict[str, SchemaInfo]:
    objects: dict[str, SchemaInfo] = {}
    if is_relevant_object(schema):
        objects[schema.refname] = schema

    if recursion_depth < 10:
        it = chain(
            schema.properties.values(),
            schema.anyOf,
            [schema.child(schema.items)] if schema.items else (),
        )
        for prop_info in it:
            objects.update(
                get_all_objects(prop_info, recursion_depth=recursion_depth + 1)
            )
    return objects


all_objects = get_all_objects(config)
all_objects
```

</details>

I didn't get to the recursive part yet, but that filter works